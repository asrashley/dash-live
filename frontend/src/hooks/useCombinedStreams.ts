import { createContext } from "preact";
import { Signal, useComputed } from "@preact/signals";

import { useAllStreams } from "./useAllStreams";
import { useAllMultiPeriodStreams } from "./useAllMultiPeriodStreams";

export type CombinedStream = {
  title: string;
  value: string;
  mps: boolean;
};

export interface UseCombinedStreamsHook {
  streamNames: Signal<string[]>;
  streamsMap: Signal<Map<string, CombinedStream>>;
  loaded: Signal<boolean>;
  error: Signal<string | null>;
}

export const UseCombinedStreams = createContext(null);

export function useCombinedStreams(): UseCombinedStreamsHook {
  const { allStreams: standardStreams, loaded: streamsLoaded, error: streamsError } = useAllStreams();
  const { streams: mpsStreams, loaded: mpsLoaded, error: mpsError } = useAllMultiPeriodStreams();
  const streamsMap = useComputed<Map<string, CombinedStream>>(() => {
    const smap = new Map();
    standardStreams.value.forEach((st) => {
      const { directory, title } = st;
      smap.set(`std.${directory}`, { title, value: directory, mps: false });
    });
    mpsStreams.value.forEach(({ name, title }) => {
      smap.set(`mps.${name}`, { title, value: name, mps: true });
    });
    return smap;
  });
  const streamNames = useComputed<string[]>(() => {
    const smap = streamsMap.value;
    const names = [...smap.keys()];
    names.sort((a, b) => {
      const t1 = smap.get(a).title;
      const t2 = smap.get(b).title;
      if (t1 == t2) {
        return 0;
      }
      if (t1 < t2) {
        return -1;
      }
      return 1;
    });
    return names;
  });
  const loaded = useComputed<boolean>(() => streamsLoaded.value && mpsLoaded.value);
  const error = useComputed<string | null>(() => {
    if (!streamsError && !mpsError) {
      return null;
    }
    return [streamsError ?? '', mpsError ?? ''].join(' ');
  })
  return { streamNames, streamsMap, loaded, error };
}
