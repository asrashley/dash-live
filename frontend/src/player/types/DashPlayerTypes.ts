export type DashPlayerTypes = 'native' | 'shaka' | 'dashjs';