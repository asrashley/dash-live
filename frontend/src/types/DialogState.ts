import { DecoratedStream } from "./DecoratedStream";

export interface ConfirmDeleteDialogState {
    name: string;
    confirmed: boolean;
}

export interface MpsOptionsDialogState {
    options: object;
    lastModified: number;
    name: string;
}

export interface TrackPickerDialogState {
    pk: number | string;
    pid: string;
    guest: boolean;
    stream: DecoratedStream;
}

export interface DialogState {
    backdrop: boolean;
    confirmDelete?: ConfirmDeleteDialogState;
    mpsOptions?: MpsOptionsDialogState;
    trackPicker?: TrackPickerDialogState;
}