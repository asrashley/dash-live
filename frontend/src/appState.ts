import { createContext } from 'preact';
import { signal, computed, type Signal, type ReadonlySignal } from "@preact/signals";
import { DialogState } from './types/DialogState';
import { PlayerLibraryState } from './types/PlayerLibraryState';

export interface AppStateType {
  backdrop: ReadonlySignal<boolean>;
  cinemaMode: Signal<boolean>;
  dialog: Signal<DialogState | null>;
  playerLibrary: Signal<PlayerLibraryState | null>;
  closeDialog: () => void;
}

export const AppStateContext = createContext<AppStateType>(null);

export function createAppState(): AppStateType {
  const dialog = signal<DialogState | null>(null);
  const cinemaMode = signal<boolean>(false);
  const playerLibrary = signal<PlayerLibraryState | null>(null);

  const backdrop = computed(() => {
    return dialog.value?.backdrop === true;
  });

  const closeDialog = () => {
    dialog.value = null;
  };

  return { cinemaMode, dialog, backdrop, playerLibrary, closeDialog };
}
