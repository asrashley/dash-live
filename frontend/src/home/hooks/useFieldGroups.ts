import { useContext } from "preact/hooks";
import { type ReadonlySignal, useComputed } from "@preact/signals";

import { fieldGroups, drmSystems } from '@dashlive/options';

import { useAllManifests  } from '../../hooks/useAllManifests';
import { UseCombinedStreams  } from '../../hooks/useCombinedStreams';
import { StaticInputProps } from "../../form/types/StaticInputProps";
import { InputFormGroup } from "../../form/types/InputFormGroup";

const drmSkipNames = new RegExp(`^${drmSystems.map(name => `${name}__enabled`).join('|')}$`);

interface UseFieldGroupsHook {
  homeFieldGroups: ReadonlySignal<InputFormGroup[]>;
}

export function useFieldGroups(): UseFieldGroupsHook {
  const { allManifests, names } = useAllManifests();
  const { streamNames, streamsMap } = useContext(UseCombinedStreams);

  const homeFieldGroups = useComputed<InputFormGroup[]>(() => {
    const playbackMode: StaticInputProps = {
      name: "mode",
      shortName: "mode",
      fullName: "playbackMode",
      title: "Playback Mode",
      type: "radio",
      options: [
        {
          title: "Video On Demand (using live profile)",
          value: "vod",
        },
        {
          title: "Live stream (using live profile)",
          value: "live",
        },
        {
          title: "Video On Demand (using on-demand profile)",
          value: "odvod",
        },
      ],
    };
    const selectManifest: StaticInputProps = {
      name: "manifest",
      shortName: "manifest",
      fullName: "manifest",
      title: "Manifest",
      text: "Manifest template to use",
      type: "select",
      options: names.value.map((name: string) => {
        const msft = allManifests.value[name];
        return {
          title: msft.title,
          value: name,
        };
      }),
    };
    const selectStream: StaticInputProps = {
      name: "stream",
      shortName: "stream",
      fullName: "stream",
      title: "Stream",
      text: "Stream to play",
      type: "select",
      options: streamNames.value.map((value: string) => ({
        title: streamsMap.value.get(value).title,
        value,
      })),
    };
    const selectDrmSystem: StaticInputProps = {
      name: "drms",
      shortName: "drms",
      fullName: "drms",
      title: "DRM systems",
      text: "DRM systems to enable for encrypted AdaptationSets",
      type: "multiselect",
      options: drmSystems.map((name: string) => ({
        name,
        title: name,
        value: name,
      })),
    };
    const generalOptions: InputFormGroup = {
      ...fieldGroups[0],
      fields: [
        playbackMode,
        selectManifest,
        selectStream,
        selectDrmSystem,
        ...fieldGroups[0].fields,
      ],
    };
    const otherGroups: InputFormGroup[] = fieldGroups.slice(1).map(grp => ({
        ...grp,
        fields: grp.fields.filter(({name}) => !drmSkipNames.test(name)),
    }));
    return [generalOptions, ...otherGroups];
  });

  return { homeFieldGroups };
}
