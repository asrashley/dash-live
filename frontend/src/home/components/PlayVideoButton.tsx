import { type ReadonlySignal, useComputed } from "@preact/signals";
import { useCallback, useContext } from "preact/hooks";
import { useLocation } from "wouter-preact";

import { CombinedStream } from "../../hooks/useCombinedStreams";
import { DashPlayerTypes } from "../../player/types/DashPlayerTypes";
import { AppStateContext } from "../../appState";
import { Icon } from "../../components/Icon";

export interface PlayVideoButtonProps {
  videoUrl: ReadonlySignal<URL>;
  stream: ReadonlySignal<CombinedStream>;
}

export function PlayVideoButton({ stream, videoUrl }: PlayVideoButtonProps) {
  const [, setLocation] = useLocation();
  const playerName = useComputed<DashPlayerTypes>(() => {
    return (videoUrl.value.searchParams.get("player") ??
      "native") as DashPlayerTypes;
  });
  const playerVersion = useComputed<string>(
    () => videoUrl.value.searchParams.get(playerName.value) ?? ""
  );
  const { playerLibrary } = useContext(AppStateContext);
  const needsReload = useComputed<boolean>(() => {
    if (playerLibrary.value === null || playerName.value === "native") {
      return false;
    }
    const { name, version } = playerLibrary.value;
    return playerName.value !== name || version !== playerVersion.value;
  });

  const onClick = useCallback(
    (ev: Event) => {
      ev.preventDefault();
      const mustReload = needsReload.value;
      playerLibrary.value = {
        name: playerName.value,
        version: playerVersion.value,
      };
      if (mustReload) {
        window.location.replace(videoUrl.value.href);
      } else {
        setLocation(videoUrl.value.href);
      }
    },
    [
      needsReload,
      playerLibrary,
      playerName,
      playerVersion,
      setLocation,
      videoUrl,
    ]
  );

  return (
    <a
      className="btn btn-lg btn-primary"
      href={videoUrl.value.href}
      data-testid="play-video-button"
      onClick={onClick}>
      <Icon name="play-fill" />
      <span className="title ps-2">Play {stream.value.title}</span>
    </a>
  );
}
