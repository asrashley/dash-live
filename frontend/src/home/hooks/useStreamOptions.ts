import { useCallback } from "preact/hooks";
import { type ReadonlySignal, type Signal, useComputed, useSignal } from "@preact/signals";

import { defaultCgiOptions, drmSystems } from "@dashlive/options";
import { CombinedStream } from "../../hooks/useCombinedStreams";
import { InputFormData } from "../../types/InputFormData";

const keyName = "dashlive.homepage.options";

function getDefaultOptions(): InputFormData {
  const lsKey = localStorage.getItem(keyName);
  const previousOptions = lsKey ? JSON.parse(lsKey) : {};
  return {
    ...defaultCgiOptions,
    manifest: "hand_made.mpd",
    mode: "vod",
    stream: undefined,
    ...previousOptions,
  };
}

const skipKeys = new RegExp(`^(${[
  ...drmSystems,
  ...drmSystems.map(name => `${name}__enabled`),
  ...drmSystems.map(name => `${name}__drmloc`),
  'drms', 'manifest', 'stream', 'mode'].join('|')})$`);

const manifestSkipKeys = /^(player|dashjs|shaka)/;

export type EnabledDrmSystems = {
  [name: string]: boolean;
};

export interface UseStreamOptionsHook {
  data: Signal<InputFormData>;
  drms: ReadonlySignal<EnabledDrmSystems>;
  stream: ReadonlySignal<CombinedStream>;
  mode: ReadonlySignal<string>;
  manifest: ReadonlySignal<string>;
  nonDefaultOptions: ReadonlySignal<object>;
  manifestOptions: ReadonlySignal<object>;
  setValue: (name: string, value: string | number | boolean) => void;
  resetAllValues: () => void;
}

const emptyStream: CombinedStream = {
  title: "",
  value: "",
  mps: false
}

export interface UseStreamOptionsProps {
  streamNames: ReadonlySignal<string[]>;
  streamsMap: ReadonlySignal<Map<string, CombinedStream>>;
}
export function useStreamOptions({ streamNames, streamsMap }: UseStreamOptionsProps): UseStreamOptionsHook {
  const data = useSignal<InputFormData>(getDefaultOptions());
  const stream = useComputed<CombinedStream>(() => {
    const name: string = (data.value['stream'] as string | undefined) ?? streamNames.value[0];
    return streamsMap.value.get(name) ?? emptyStream;
  });
  const mode = useComputed<string>(() => data.value['mode'] as string);
  const manifest = useComputed<string>(() => data.value['manifest'] as string);
  const drms = useComputed<EnabledDrmSystems>(() => Object.fromEntries(drmSystems.map(name => [name, data.value[name] === "1"])));
  const nonDefaultOptions = useComputed<object>(() => {
    const params = Object.entries(data.value)
      .filter(([key, value]) => defaultCgiOptions[key] != value)
      .filter(([key]) => !skipKeys.test(key));
    const drm = drmSystems.filter(system => data.value[system] === "1").map(system => {
      const drmLoc = data.value[`${system}__drmloc`];
      return drmLoc ? `${system}-${drmLoc}`: system;
    });
    if (drm.length) {
      params.push(['drm', drm.join(',')]);
    }
    return Object.fromEntries(params);
  });
  const manifestOptions = useComputed<object>(() => Object.fromEntries(
    Object.entries(nonDefaultOptions.value).filter(([key]) => !manifestSkipKeys.test(key))));

  const setValue = useCallback(
    (name: string, value: string | number | boolean) => {
      if (value === true) {
        value = "1";
      } else if (value === false) {
        value = "0";
      }
      data.value = {
        ...data.value,
        [name]: value,
      };
      if (name === "stream" && mode.value === "odvod") {
        const nextStream = streamsMap.value.get(value as string);
        if (nextStream?.mps) {
          data.value = {
            ...data.value,
            mode: "vod",
          };
        }
      }
      const params = Object.fromEntries(
        Object.entries(data.value).filter(
          ([key, value]) => defaultCgiOptions[key] !== value
        )
      );
      localStorage.setItem(keyName, JSON.stringify(params));
    },
    [data, mode, streamsMap]
  );

  const resetAllValues = useCallback(() => {
    data.value = { ...getDefaultOptions() };
    localStorage.removeItem(keyName);
  }, [data]);

  return { data, drms, stream, mode, manifest, nonDefaultOptions, manifestOptions, setValue, resetAllValues };
}
