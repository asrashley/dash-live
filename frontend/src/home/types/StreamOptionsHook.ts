import { createContext } from "preact";
import { UseStreamOptionsHook } from "../hooks/useStreamOptions";

export const StreamOptionsContext = createContext<UseStreamOptionsHook>(null);
