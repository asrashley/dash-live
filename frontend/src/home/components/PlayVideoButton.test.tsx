import {
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  test,
  vi,
} from "vitest";
import { fireEvent } from "@testing-library/preact";
import { signal } from "@preact/signals";
import { useLocation } from "wouter-preact";

import { CombinedStream } from "../../hooks/useCombinedStreams";
import { renderWithProviders } from "../../test/renderWithProviders";
import { PlayVideoButton } from "./PlayVideoButton";
import { DashPlayerTypes } from "../../player/types/DashPlayerTypes";
import { AppStateType } from "../../appState";
import { DialogState } from "../../types/DialogState";
import { PlayerLibraryState } from "../../types/PlayerLibraryState";

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
  };
});

describe("PlayVideoButton component", () => {
  const useLocationSpy = vi.mocked(useLocation);
  const setLocation = vi.fn();
  const mockNavigate = vi.fn();
  const videoUrl = signal<URL>(new URL("http://example.local"));
  const stream = signal<CombinedStream>({
    title: "Stream Title",
    value: "bbb",
    mps: false,
  });
  const dialog = signal<DialogState | null>(null);
  const cinemaMode = signal<boolean>(false);
  const playerLibrary = signal<PlayerLibraryState | null>(null);
  const mockLocation = {
    ...new URL(document.location.href),
    pathname: "/",
    replace: vi.fn(),
  };

  function setVideoUrl(player: DashPlayerTypes, version: string) {
    const url = `http://example.local?player=${player}&${player}=${version}`;
    videoUrl.value = new URL(url);
  }

  beforeAll(() => {
    vi.stubGlobal("location", mockLocation);
  });

  beforeEach(() => {
    useLocationSpy.mockReturnValue(["/play", setLocation]);
    playerLibrary.value = null;
    dialog.value = null;
    cinemaMode.value = false;
    vi.stubGlobal("navigation", mockNavigate);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test.each<DashPlayerTypes>(["dashjs", "shaka", "native"])(
    "matches snapshot for %s",
    (player: DashPlayerTypes) => {
      setVideoUrl(player, "1.0.0");
      const { asFragment } = renderWithProviders(
        <PlayVideoButton stream={stream} videoUrl={videoUrl} />
      );
      expect(asFragment()).toMatchSnapshot();
    }
  );

  test("navigates to video URL without reload", () => {
    setVideoUrl("native", "");
    const { getByTestId, appState } = renderWithProviders(
      <PlayVideoButton stream={stream} videoUrl={videoUrl} />
    );
    const button = getByTestId("play-video-button") as HTMLAnchorElement;
    fireEvent.click(button);
    expect(setLocation).toHaveBeenCalledWith(videoUrl.value.href);
    expect(mockLocation.replace).not.toHaveBeenCalled();
    expect(appState.playerLibrary.value).toEqual({
      name: "native",
      version: "",
    });
  });

  test("switching to native player does not need a reload", () => {
    playerLibrary.value = {
      name: "dashjs",
      version: "1.0.0",
    };
    setVideoUrl("native", "");
    const { getByTestId, appState } = renderWithProviders(
      <PlayVideoButton stream={stream} videoUrl={videoUrl} />
    );
    const button = getByTestId("play-video-button") as HTMLAnchorElement;
    fireEvent.click(button);
    expect(setLocation).toHaveBeenCalledWith(videoUrl.value.href);
    expect(mockLocation.replace).not.toHaveBeenCalled();
    expect(appState.playerLibrary.value).toEqual({
      name: "native",
      version: "",
    });
  });

  test("reloads when player changes", () => {
    const initialAppState: AppStateType = {
      dialog,
      cinemaMode,
      playerLibrary,
      backdrop: undefined,
      closeDialog: vi.fn,
    };
    playerLibrary.value = {
      name: "dashjs",
      version: "1.0.0",
    };
    setVideoUrl("shaka", "1.2.3");
    const { getByTestId, appState } = renderWithProviders(
      <PlayVideoButton stream={stream} videoUrl={videoUrl} />,
      { appState: initialAppState }
    );
    expect(appState.playerLibrary.value).toEqual(playerLibrary.value);
    const button = getByTestId("play-video-button") as HTMLAnchorElement;
    fireEvent.click(button);
    expect(setLocation).not.toHaveBeenCalled();
    expect(mockLocation.replace).toHaveBeenCalledWith(videoUrl.value.href);
  });
});
