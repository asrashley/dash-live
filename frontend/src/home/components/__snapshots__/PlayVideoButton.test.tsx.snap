// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayVideoButton component > matches snapshot for dashjs 1`] = `
<DocumentFragment>
  <a
    class="btn btn-lg btn-primary"
    data-testid="play-video-button"
    href="http://example.local/?player=dashjs&dashjs=1.0.0"
  >
    <span
      aria-label="play-fill icon"
      class="bi icon bi-play-fill "
      role="img"
    />
    <span
      class="title ps-2"
    >
      Play Stream Title
    </span>
  </a>
</DocumentFragment>
`;

exports[`PlayVideoButton component > matches snapshot for native 1`] = `
<DocumentFragment>
  <a
    class="btn btn-lg btn-primary"
    data-testid="play-video-button"
    href="http://example.local/?player=native&native=1.0.0"
  >
    <span
      aria-label="play-fill icon"
      class="bi icon bi-play-fill "
      role="img"
    />
    <span
      class="title ps-2"
    >
      Play Stream Title
    </span>
  </a>
</DocumentFragment>
`;

exports[`PlayVideoButton component > matches snapshot for shaka 1`] = `
<DocumentFragment>
  <a
    class="btn btn-lg btn-primary"
    data-testid="play-video-button"
    href="http://example.local/?player=shaka&shaka=1.0.0"
  >
    <span
      aria-label="play-fill icon"
      class="bi icon bi-play-fill "
      role="img"
    />
    <span
      class="title ps-2"
    >
      Play Stream Title
    </span>
  </a>
</DocumentFragment>
`;
